rfcs/0140-simple-package-paths.md

```diff
@@ -0,0 +1,234 @@
+
+## Custom arguments
+
+If `pkgs/top-level/all-packages.nix` contains a definition for the attribute `${name}` and the unit directory `pkgs/unit/${shard}/${name}` exists, then the attribute value must be defined as `pkgs.callPackage pkgs/unit/${shard}/${name}/pkg-fun.nix args`, where `args` may be a freely chosen expression.
+
+This ensures that even if a package initially doesn't require a custom `args`, if it later does, it doesn't have to be moved out of the `pkgs/unit` directory to pass custom arguments.
+
+## Examples
+[examples]: #examples
+
+To add a new package `pkgs.foobar` to Nixpkgs, one only needs to create the file `pkgs/unit/fo/foobar/pkg-fun.nix`.
+No need to find an appropriate category nor to modify `pkgs/top-level/all-packages.nix` anymore.
+
+With many packages, the `pkgs/unit` directory may look like this:
+
+```
+pkgs
+└── unit
+   ├── _0
+   │  ├── _0verkill
+   │  └── _0x
+   ┊
+   ├── ch
+   │  ├── ChowPhaser
+   │  ├── CHOWTapeModel
+   │  ├── chroma
+   │  ┊
+   ┊
+   ├── t
+   │  └── t
+   ┊
+```
+
+# Interactions
+[interactions]: #interactions
+
+- `nix edit` and search.nixos.org are unaffected, since they rely on `meta.position` to get the file to edit, which still works
+- `git blame` locally and on GitHub is unaffected, since it follows file renames properly.
+- A commonly recommended way of building package directories in Nixpkgs is to use `nix-build -E 'with import <nixpkgs> {}; callPackage pkgs/applications/misc/hello {}'`.
+  Since the path changes `pkg-fun.nix` is now used, this becomes like `nix-build -E 'with import <nixpkgs> {}; callPackage pkgs/unit/he/hello/pkg-fun.nix {}'`, which is harder for users.
+  However, calling a path like this is an anti-pattern anyway, because it doesn't use the correct Nixpkgs version and it doesn't use the correct argument overrides.
+  The correct way of doing it was to add the package to `pkgs/top-level/all-packages.nix`, then calling `nix-build -A hello`.
+  This `nix-build -E` workaround is partially motivated by the difficulty of knowing the mapping from attributes to package paths, which is what this RFC improves upon.
+  By teaching users that `pkgs/unit/*/<name>` corresponds to `nix-build -A <name>`, the need for such `nix-build -E` workarounds should disappear.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- The existing categorization of packages gets lost. Counter-arguments:
+  - It was never that useful to begin with
+    - The categorization was always incomplete, because packages defined in the language package sets often don't get their own categorized file path.
+    - It was an inconvenient user interface, requiring a checkout or browsing through GitHub
+    - Many packages fit multiple categories, leading to multiple locations to search through instead of one
+  - There's other better ways of discovering similar packages, e.g. [Repology](https://repology.org/)
```

Comment:
How would you do that?

For example in https://github.com/NixOS/nixpkgs/tree/master/pkgs/games i see nix packages related to "games". How would i see such a list on Repology?

I think having a category as metadata for a package would be a good future improvement. We can for example extract it from `menu-spec` of Freedesktop.org Specifications (https://specifications.freedesktop.org/menu-spec/latest/apa.html) and show it in https://search.nixos.org/ and use for `nix search`.

Not sure if the current categories have any value and should be preserved. But we have them in git history, so they could get extracted after a change.

Reply:
As you say, these categories only have value in certain contexts. The Nix language is much more flexible for providing that functionality, and all of that can easily move into `meta`.